### ISSUE TITLE
Break line when over max_line_length

### ISSUE THREAD

**Issue by rodgavila** [pinterest/ktlint#251]
Ktlint currently has a rule for "lines to long", however it cannot auto-fix it. This I believe is the only rule that cannot be autofixed, and by implementing would guarantee that the linter would always succeed after -format is passed.

**Comment by gortiz** [comment]
I think wildcard import is another rule that cannot be autofixed

**Comment by cable729** [comment]
Is this being worked on?

**Comment by shyiko** [comment]
@cable729 Not directly. I expect this to change once https://github.com/shyiko/ktlint/pull/294 is merged, though.

**Comment by cable729** [comment]
Any updates here?

**Comment by hravichandran0** [comment]
This is one of the few major rules of ktlint that are checked for but cannot be auto-fixed. There have been no updates for a year.

"Not directly. I expect this to change once #294 is merged, though."

#294 was closed on Feb 11 2019. Any idea the timeline on this ticket?

**Comment by sonahovhannisyan** [comment]
@Tapchicoma, I see you added this as an enhancement. Might you also have an approximate idea when the auto-format for this  will be implemented?

**Comment by Tapchicoma** [comment]
@sonahovhannisyan no idea, PRs are welcome :slightly_smiling_face:

**Comment by cdavietei** [comment]
I've been trying to look into a fix for this, and I think that the existing [IndentationRule](https://github.com/pinterest/ktlint/blob/master/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/IndentationRule.kt) can be re-used in some way as that already handles line breaking and indentation, see this [test case](https://github.com/pinterest/ktlint/blob/master/ktlint-ruleset-standard/src/test/resources/spec/indent/format-binary-expression.kt.spec) and it's [expected value](https://github.com/pinterest/ktlint/blob/master/ktlint-ruleset-standard/src/test/resources/spec/indent/format-binary-expression-expected.kt.spec).

The options as I see it are:
1. Integrate the [MaxLineLengthRule](https://github.com/pinterest/ktlint/blob/master/ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/MaxLineLengthRule.kt) into IndentationRule
2. Extract the formatting code (specifically the knowledge of when to break the line) from IndentationRule into a shared class for use in both Rules

There is another option which is to implement the logic for where to break functions, lines, and classes according to the requirements outline by the [Android style guide](https://developer.android.com/kotlin/style-guide#where_to_break) and the various break rules outlined throughout [Kotlin's code conventions doc](https://kotlinlang.org/docs/coding-conventions.html). I wouldn't want to do it this way as we'd end up with a lot of potentially duplicate code.

I'm not super familiar with this codebase but am willing to make the change and PR if anyone else can validate this hunch or give insight the options I've presented or others I might be missing.

**Comment by toddobryan** [comment]
To make matters worse, ktlint seems to be causing the problem and then complaining that it can't fix it.

I'm using KotlinPoet to generate some code. The code that comes out isn't formatted so well, so I've been trying to run ktlint afterward, so that if anybody actually has to look at the code, it's easy to read.

But I'm getting a problem because KotlinPoet outputs an extension function with the `=` on the next line. It has to do this because putting it on the same line as the declaration would exceed the 100 character max line length.

When I run `ktlint -F` on this file, it moves the `=` up a line and then promptly fails because it can't fix the too long line it just created. This fails the whole build.

I mean it's one thing to say we don't support auto-fixing max line length errors, but when you take a file where all the lines are under the max line length and then produce a new file that has lines over the max line length, that seems like a bug.

**Comment by paul-dingemans** [comment]
> To make matters worse, ktlint seems to be causing the problem and then complaining that it can't fix it.
> 
> I'm using KotlinPoet to generate some code. The code that comes out isn't formatted so well, so I've been trying to run ktlint afterward, so that if anybody actually has to look at the code, it's easy to read.
> 
> But I'm getting a problem because KotlinPoet outputs an extension function with the `=` on the next line. It has to do this because putting it on the same line as the declaration would exceed the 100 character max line length.
> 
> When I run `ktlint -F` on this file, it moves the `=` up a line and then promptly fails because it can't fix the too long line it just created. This fails the whole build.
> 
> I mean it's one thing to say we don't support auto-fixing max line length errors, but when you take a file where all the lines are under the max line length and then produce a new file that has lines over the max line length, that seems like a bug.

This indeed sound like a bug. It should however not be related to this current issue as it is not very likely to be picked up. Please create a new issue and post your concrete example with `.editorconfig` settings.

**Comment by xenoterracide** [comment]
same issue, I'll create a new bug...

**Comment by paul-dingemans** [comment]
Any idea's to break up long lines should be broken down in smaller issues and preferably be part of other rules. For example the `function-signature` avoids long lines by taking the max_line_length parameter into account.
